cpp/ParquetFileWriter.cpp

```diff
@@ -90,8 +90,8 @@ extern "C"
 	{
 		TRYCATCH
 		(
-			const auto m = writer->key_value_metadata();
-			*key_value_metadata = m ? new std::shared_ptr<const KeyValueMetadata>(m) : nullptr;
+			const auto& m = writer->key_value_metadata();
+			*key_value_metadata = m ? new std::shared_ptr(m) : nullptr;
```

Comment:
I don't fully grasp what changing `const auto` to `const auto&` does, and why it's not done on line 102 below. Could you please explain? 😅

Reply:
Probably one that was highlighted by ReSharper C++, as it's quite subtle.

`const auto&` means "a constant reference", whereas `const auto` implies a constant copy. For some reason, `writer->key_value_metadata()` returns a reference to a smart pointer, and `writer->metadata()` returns a smart pointer copy. 

Thus slight inconsistency within Arrow here (probably doesn't matter), but at least now our code correctly reflects this difference.